File: CBLClient/Apps/JavaRequestHandlers/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/ReplicatorRequestHandler.java
package com.couchbase.mobiletestkit.javacommon.RequestHandler;

import java.util.ArrayList;
import java.util.List;

import com.couchbase.mobiletestkit.javacommon.Args;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.DocumentReplication;
import com.couchbase.lite.DocumentReplicationListener;
import com.couchbase.lite.ListenerToken;
import com.couchbase.lite.ReplicatedDocument;
import com.couchbase.lite.Replicator;
import com.couchbase.lite.ReplicatorChange;
import com.couchbase.lite.ReplicatorChangeListener;
import com.couchbase.lite.ReplicatorConfiguration;
import com.couchbase.lite.ReplicatorStatus;
import com.couchbase.mobiletestkit.javacommon.util.ConcurrentExecutor;
import com.couchbase.CouchbaseLiteServ.util.Log;


public class ReplicatorRequestHandler {
    private static final String TAG = "Replicator";
    /* -------------- */
    /* - Replicator - */
    /* -------------- */

    public Replicator create(Args args) {
        ReplicatorConfiguration config = args.get("config", ReplicatorConfiguration.class);
        return new Replicator(config);
    }

    public String status(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        return replicator.getStatus().toString();
    }

    public String getActivityLevel(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        return replicator.getStatus().getActivityLevel().toString().toLowerCase();
    }

    public ReplicatorChangeListener addChangeListener(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        MyReplicatorListener changeListener = new MyReplicatorListener();
        ListenerToken token = replicator.addChangeListener(ConcurrentExecutor.EXECUTOR, changeListener);
        changeListener.setToken(token);
        return changeListener;
    }

    public void removeChangeListener(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        MyReplicatorListener changeListener = args.get("changeListener", MyReplicatorListener.class);
        replicator.removeChangeListener(changeListener.getToken());
    }

    public MyDocumentReplicatorListener addReplicatorEventChangeListener(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        MyDocumentReplicatorListener changeListener = new MyDocumentReplicatorListener();
        ListenerToken token = replicator.addDocumentReplicationListener(ConcurrentExecutor.EXECUTOR, changeListener);
        changeListener.setToken(token);
        return changeListener;
    }

    public void removeReplicatorEventListener(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        MyDocumentReplicatorListener changeListener = args.get("changeListener", MyDocumentReplicatorListener.class);
        replicator.removeChangeListener(changeListener.getToken());
    }

    public int changeListenerChangesCount(Args args) {
        MyReplicatorListener changeListener = args.get("changeListener", MyReplicatorListener.class);
        return changeListener.getChanges().size();
    }

    public List<String> replicatorEventGetChanges(Args args) {
        MyDocumentReplicatorListener changeListener = args.get("changeListener", MyDocumentReplicatorListener.class);
        List<DocumentReplication> changes = changeListener.getChanges();
        List<String> event_list = new ArrayList<>();

        for (DocumentReplication change : changes) {
            for (ReplicatedDocument document : change.getDocuments()) {
                String doc_id = "doc_id: " + document.getID();
                String error = ", error_code: ";
                String error_domain = "0";
                int error_code = 0;

                if (document.getError() != null) {
                    error_code = document.getError().getCode();
                    error_domain = document.getError().getDomain();
                }
                error = error + error_code + ", error_domain: " + error_domain;
                String flags = ", flags: " + document.getFlags();
                String push = ", push: " + change.isPush();
                event_list.add(doc_id + error + push + flags);
            }
        }
        return event_list;
    }

    public String toString(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        return replicator.toString();
    }

    public void start(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        replicator.start();
    }

    public void stop(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        replicator.stop();
    }

    public Replicator changeGetReplicator(Args args) {
        ReplicatorChange change = args.get("change", ReplicatorChange.class);
        return change.getReplicator();
    }

    public ReplicatorStatus changeGetStatus(Args args) {
        ReplicatorChange change = args.get("change", ReplicatorChange.class);
        return change.getStatus();
    }

    public int replicatorEventChangesCount(Args args) {
        MyDocumentReplicatorListener changeListener = args.get("changeListener", MyDocumentReplicatorListener.class);
        return changeListener.getChanges().size();
    }

    public List<ReplicatorChange> changeListenerGetChanges(Args args) {
        MyReplicatorListener changeListener = args.get("changeListener", MyReplicatorListener.class);
        return changeListener.getChanges();
    }

    public CouchbaseLiteException replicatorGetError(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        Log.i(TAG, "Replication Error... ");
        Log.i(TAG, "replicator.getStatus().getError()");
        return replicator.getStatus().getError();
    }

    public ReplicatorConfiguration config(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        return replicator.getConfig();
    }

    public long getCompleted(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        return replicator.getStatus().getProgress().getCompleted();
    }

    public long getTotal(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        return replicator.getStatus().getProgress().getTotal();
    }

    public String getError(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        CouchbaseLiteException error = replicator.getStatus().getError();
        if (error != null) {
            return error.toString();
        }
        return null;
    }

    public Boolean isContinuous(Args args) {
        ReplicatorConfiguration config = args.get("config", ReplicatorConfiguration.class);
        return config.isContinuous();
    }

    public void resetCheckpoint(Args args) {
        Replicator replicator = args.get("replicator", Replicator.class);
        replicator.start(true);
    }

}

class MyReplicatorListener implements ReplicatorChangeListener {
    private final List<ReplicatorChange> changes = new ArrayList<>();
    private ListenerToken token;

    public List<ReplicatorChange> getChanges() {
        return changes;
    }

    public void setToken(ListenerToken token) {
        this.token = token;
    }

    public ListenerToken getToken() {
        return token;
    }

    @Override
    public void changed(ReplicatorChange change) {
        changes.add(change);
    }
}

class MyDocumentReplicatorListener implements DocumentReplicationListener {
    private final List<DocumentReplication> changes = new ArrayList<>();
    private ListenerToken token;

    public List<DocumentReplication> getChanges() {
        return changes;
    }

    public void setToken(ListenerToken token) {
        this.token = token;
    }

    public ListenerToken getToken() {
        return token;
    }

    @Override
    public void replication(DocumentReplication replication) {
        changes.add(replication);
    }
}

File: CBLClient/Apps/JavaRequestHandlersKotlin/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/ReplicatorRequestHandler.kt
package com.couchbase.mobiletestkit.javacommon.RequestHandler

import com.couchbase.lite.*
import com.couchbase.mobiletestkit.javacommon.Args
import com.couchbase.mobiletestkit.javacommon.util.ConcurrentExecutor
import com.couchbase.mobiletestkit.javacommon.util.Log
import java.util.*

class ReplicatorRequestHandler {
    /* -------------- */ /* - Replicator - */ /* -------------- */
    fun create(args: Args): Replicator {
        val config = args.get<ReplicatorConfiguration>("config")
        return Replicator(config)
    }

    fun getConfig(args: Args): ReplicatorConfiguration {
        val replicator = args.get<Replicator>("replicator")
        return replicator.config
    }

    fun status(args: Args): String {
        val replicator = args.get<Replicator>("replicator")
        return replicator.status.toString()
    }

    fun getActivityLevel(args: Args): String {
        val replicator = args.get<Replicator>("replicator")
        return replicator.status.activityLevel.toString().toLowerCase()
    }

    fun addChangeListener(args: Args): ReplicatorChangeListener {
        val replicator = args.get<Replicator>("replicator")
        val changeListener = MyReplicatorListener()
        val token = replicator.addChangeListener(ConcurrentExecutor.EXECUTOR, changeListener)
        changeListener.token = token
        return changeListener
    }

    fun removeChangeListener(args: Args) {
        val replicator = args.get<Replicator>("replicator")
        val changeListener: MyReplicatorListener = args["changeListener"]
        replicator.removeChangeListener(changeListener.token!!)
    }

    fun addReplicatorEventChangeListener(args: Args): MyDocumentReplicatorListener {
        val replicator = args.get<Replicator>("replicator")
        val changeListener = MyDocumentReplicatorListener()
        val token =
            replicator.addDocumentReplicationListener(ConcurrentExecutor.EXECUTOR, changeListener)
        changeListener.token = token
        return changeListener
    }

    fun removeReplicatorEventListener(args: Args) {
        val replicator = args.get<Replicator>("replicator")
        val changeListener: MyDocumentReplicatorListener = args["changeListener"]
        replicator.removeChangeListener(changeListener.token!!)
    }

    fun changeListenerChangesCount(args: Args): Int {
        val changeListener: MyReplicatorListener = args["changeListener"]
        return changeListener.getChanges().size
    }

    fun replicatorEventGetChanges(args: Args): List<String> {
        val changeListener: MyDocumentReplicatorListener = args["changeListener"]
        val changes = changeListener.getChanges()
        val event_list: MutableList<String> = ArrayList()
        for (change in changes) {
            for (document in change.documents) {
                val event = document.toString()
                val doc_id = "doc_id: " + document.id
                var error = ", error_code: "
                var error_domain = "0"
                var error_code = 0
                if (document.error != null) {
                    error_code = document.error!!.code
                    error_domain = document.error!!.domain
                }
                error = "$error$error_code, error_domain: $error_domain"
                val flags = ", flags: " + document.getFlags()
                val push = ", push: " + change.isPush
                event_list.add(doc_id + error + push + flags)
            }
        }
        return event_list
    }

    fun toString(args: Args): String {
        val replicator = args.get<Replicator>("replicator")
        return replicator.toString()
    }

    fun start(args: Args) {
        val replicator = args.get<Replicator>("replicator")
        replicator.start()
    }

    fun stop(args: Args) {
        val replicator = args.get<Replicator>("replicator")
        replicator.stop()
    }

    fun changeGetReplicator(args: Args): Replicator {
        val change = args.get<ReplicatorChange>("change")
        return change.replicator
    }

    fun changeGetStatus(args: Args): ReplicatorStatus {
        val change = args.get<ReplicatorChange>("change")
        return change.status
    }

    fun replicatorEventChangesCount(args: Args): Int {
        val changeListener: MyDocumentReplicatorListener = args["changeListener"]
        return changeListener.getChanges().size
    }

    fun changeListenerGetChanges(args: Args): List<ReplicatorChange> {
        val changeListener: MyReplicatorListener = args["changeListener"]
        return changeListener.getChanges()
    }

    fun replicatorGetError(args: Args): CouchbaseLiteException? {
        val replicator = args.get<Replicator>("replicator")
        Log.i(TAG, "Replication Error... ")
        Log.i(TAG, "replicator.getStatus().getError()")
        return replicator.status.error
    }

    fun config(args: Args): ReplicatorConfiguration {
        val replicator = args.get<Replicator>("replicator")
        return replicator.config
    }

    fun getCompleted(args: Args): Long {
        val replicator = args.get<Replicator>("replicator")
        return replicator.status.progress.completed
    }

    fun getTotal(args: Args): Long {
        val replicator = args.get<Replicator>("replicator")
        return replicator.status.progress.total
    }

    fun getError(args: Args): String? {
        val replicator = args.get<Replicator>("replicator")
        val error = replicator.status.error
        return error?.toString()
    }

    fun isContinuous(args: Args): Boolean {
        val config = args.get<ReplicatorConfiguration>("config")
        return config.isContinuous
    }

    fun resetCheckpoint(args: Args) {
        val replicator = args.get<Replicator>("replicator")
        replicator.start(true)
    }

    companion object {
        private const val TAG = "Replicator"
    }
}

class MyReplicatorListener : ReplicatorChangeListener {
    private val changes: MutableList<ReplicatorChange> = ArrayList()
    var token: ListenerToken? = null

    fun getChanges(): List<ReplicatorChange> {
        return changes
    }

    override fun changed(change: ReplicatorChange) {
        changes.add(change)
    }
}

class MyDocumentReplicatorListener : DocumentReplicationListener {
    private val changes: MutableList<DocumentReplication> = ArrayList()
    var token: ListenerToken? = null

    fun getChanges(): List<DocumentReplication> {
        return changes
    }

    override fun replication(replication: DocumentReplication) {
        changes.add(replication)
    }
}